Tidy user takeoff

peterbarker opened the pull request:
The logic around the poshold changes is the most complicated to nut out.  But this seems to be correct and makes it look like the other modes.

khancyr commented:
Hummm, it seems ok. Did you try on poshold on Heli ?

OXINARF commented on ArduCopter/mode_poshold.cpp:
```diff
@@ -156,25 +156,13 @@ void Copter::ModePosHold::run()
         target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
 
+        // initiate take-off
+        if (!takeoff.running()) {
```
This is missing the call to `triggered`, it should be:
```suggestion
        if (takeoff.triggered() && !takeoff.running()) {
```

peterbarker replied:
Fixed.

peterbarker commented on ArduCopter/takeoff.cpp:
```diff
@@ -50,11 +50,20 @@ bool Copter::Mode::do_user_takeoff(float takeoff_alt_cm, bool must_navigate)
+    // indicate we are taking off
+    copter.set_land_complete(false);
+    // tell position controller to reset alt target and reset I terms
+    copter.pos_control->init_takeoff();
```
Actually, moving it to *after* those early-returns would make it a functional change!

The code that's being removed unconditionally ran these lines - since `start` is a void function they really didn't have much choice in the matter.

While I do agree that this all looks a little odd, this could actually be critical given we're resetting i-terms and the like.

OXINARF replied:
> While I do agree that this all looks a little odd, this could actually be critical given we're resetting i-terms and the like.

I don't understand this, how is it critical if the takeoff isn't running? It should be as critical as before takeoff is engaged...

bnsgeyer commented:
From what I can tell, this change looks fine for trad heli but I will want to do some simulator testing before I can say for certain.

peterbarker commented:
@khancyr autotest added

@oxinarf comment fixed.

peterbarker commented:
On Wed, 28 Nov 2018, Pierre Kancir wrote:

> should be double quote " ;-P

Fixed.


> +            self.set_rc(3, 1600)
> 1500 ? you didn't back

Fixed.

bnsgeyer commented:
@peterbarker  what was your intention of adding a takeoff in poshold? I just wanted to check a new logic to make sure there were no issues for takeoff  in that mode.

peterbarker commented:
On Wed, 28 Nov 2018, Bill Geyer wrote:

> @peterbarker what was your intention of adding a takeoff in poshold? I just wanted to check a new logic to make sure there were no issues for takeoff in
> that mode.

It was already there - it was just different-looking to all the rest of 
the modes which have it.

See the autotest to see how it is triggered - I believe this is the first 
test which tests the pilot-takeoff-alt feature.

peterbarker commented:
Now includes a fix for https://github.com/ArduPilot/ardupilot/issues/9969

rmackay9 commented:
As discussed on the dev call this will conflict badly with the spool logic PR so we need to get the spool logic PR in first (sorry!). https://github.com/ArduPilot/ardupilot/pull/10656
